master/internal/restore.go

```diff
@@ -180,3 +189,93 @@ func (e *experiment) snapshotAndSave(ctx *actor.Context, ts trialSnapshot) {
 		return
 	}
 }
+
+// experimentSnapshotShims maps a version to the shim that bumps that version.
+var experimentSnapshotShims = map[int]snapshotShimFunc{
+	0: shimExperimentSnapshotV0,
+}
+
+// shimExperimentSnapshot shims a trial snapshot to the version required by the master,
+// returning an error in the event the shim fails or the snapshot version is greater
+// than the current version (which could happen in a downgrade).
+func shimExperimentSnapshot(snapshot []byte, version int) ([]byte, error) {
+	return shimSnapshot(experimentSnapshotShims, snapshot, version)
+}
+
+// trialSnapshotShims maps a version to the shim that bumps that version.
+var trialSnapshotShims = map[int]snapshotShimFunc{
+	0: identidyShim,
+}
+
+// shimTrialSnapshot shims a trial snapshot to the version required by the master,
+// returning an error in the same cases as shimExperimentSnapshot.
+func shimTrialSnapshot(snapshot []byte, version int) ([]byte, error) {
+	return shimSnapshot(trialSnapshotShims, snapshot, version)
+}
+
+func shimSnapshot(shims map[int]snapshotShimFunc, snapshot []byte, version int) ([]byte, error) {
+	if version > experimentSnapshotVersion {
+		return nil, fmt.Errorf("cannot shim from %d to %d", experimentSnapshotVersion, version)
+	}
+	var err error
+	for version < experimentSnapshotVersion {
+		shim, ok := shims[version]
+		if !ok {
+			return nil, fmt.Errorf("missing shim from %d to %d", experimentSnapshotVersion, version)
+		}
+		if snapshot, err = shim(snapshot); err != nil {
+			return nil, errors.Wrapf(err, "failed to shim snapshot")
+		}
+		version++
+	}
+	return snapshot, nil
+}
+
+// snapshotShimFunc is a shimming function.
+type snapshotShimFunc func([]byte) ([]byte, error)
+
+// identidyShim is a shim that does nothing.
+func identidyShim(snapshot []byte) ([]byte, error) {
+	return snapshot, nil
+}
+
+// Version 0 => 1 shims
+
+func shimExperimentSnapshotV0(snapshot []byte) ([]byte, error) {
```

Comment:
Probably worth including some context in a comment about when/why we wrote this shim.  Details like:
* we used to have searcher checkpoints
* now checkpoints are always created before validations, so they don't need to be explicit
* only pbt cared about searcher checkpoints completing anyway
* this is moving us in the direction of a world where checkpoints are pushed to the searcher, rather than asked for by the searcher

Normally, it's good practice to avoid comments which are true at a specific point in time (anything that starts with "it used to be like X but now ..."), but since this shim totally corresponds to a specific point in time, I think this would be pretty helpful.